### PR TITLE
Micrometer: SimpleMeterRegistry is not created if no other registry is defined #5023

### PR DESCRIPTION
**Pull request by JiriOndrusek**
fixes https://github.com/apache/camel-quarkus/issues/5023

Quarkus gathers all registerie and puts them together as CompositeRegistry (created by [code](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/Metrics.java#L36))
This happens even if there is no metric implementation dependency (which causes thet CompositeRepository is empty and can not be used). For such purposes the Camel component creates a SimpleMeterRegistry for the testing purposes  ([code](https://github.com/apache/camel/blob/main/components/camel-micrometer/src/main/java/org/apache/camel/component/micrometer/MicrometerUtils.java#L67)).
This PR adds code, which detects cases where CompositeRegistry is empty and in such cases  creates `SimpleMeterRegistry`.

### CODE REVIEW

File: extensions/micrometer/deployment/src/main/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerProcessor.java [hunk 1]
```diff
@@ -65,4 +67,13 @@ RuntimeCamelContextCustomizerBuildItem runtimeContextCustomizer(
         return new RuntimeCamelContextCustomizerBuildItem(
                 recorder.createRuntimeContextCustomizer(config, rootMeterRegistryBuildItem.getValue()));
     }
+
+    @Record(ExecutionTime.STATIC_INIT)
+    @BuildStep
+    CamelRuntimeBeanBuildItem tikaComponent(CamelMicrometerRecorder recorder) {
```
Comment (jamesnetherton) [inline extensions/micrometer/deployment/src/main/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerProcessor.java:73]:
```suggestion
    CamelRuntimeBeanBuildItem micrometerComponent(CamelMicrometerRecorder recorder) {
```

File: extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+public class MicrometerSimpleMeterRegistryTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest CONFIG = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class));
+    @Inject
+    CamelContext context;
+
+    @Test
+    public void testQuarkusSchedulerAutowired() throws Exception {
```
Comment (jamesnetherton) [inline extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java:40]:
```suggestion
    public void testDefaultSimpleMeterRegistry() throws Exception {
```

File: extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        MicrometerComponent component = context.getComponent("micrometer", MicrometerComponent.class);
+        assertTrue(component.getMetricsRegistry() instanceof SimpleMeterRegistry, "SimpleMeterRegistry");
```
Comment (jamesnetherton) [inline extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java:42]:
```suggestion
        assertInstanceOf(SimpleMeterRegistry.class, component.getMetricsRegistry() );
```

**Comment by jamesnetherton** [comment]
I wonder if we should avoid extending the component? 

And instead consume `List<MicrometerRegistryProviderBuildItem>`. And if it's empty, then we can just register `SimpleMeterRegistry` with the existing Quarkus composite registry (which has its own `BuildItem` where the value can be passed to a recorder).

I'm just thinking out loud, not sure if that would work or not.

**Comment by JiriOndrusek** [comment]
> I wonder if we should avoid extending the component?
> 
> And instead consume `List<MicrometerRegistryProviderBuildItem>`. And if it's empty, then we can just register `SimpleMeterRegistry` with the existing Quarkus composite registry (which has its own `BuildItem` where the value can be passed to a recorder).
> 
> I'm just thinking out loud, not sure if that would work or not.

This is a nice idea. Registering simpleRegistry and keeping the Quarks mechanism working as expected should be nice and it would allow to share SimpleRegistry among more services/apps/beans. I'll try that solution.

and thanks for the founded issues in the code , I'll rework the PR with this new approach to see if it works as expected.

**Comment by JiriOndrusek** [comment]
@jamesnetherton 
my first attempt was not successful. Unfortunately it is not possible to consume  `List<MicrometerRegistryProviderBuildItem>` and create a new `MicrometerRegistryProviderBuildItem` at the same time. I'm getting `io.quarkus.builder.ChainBuildException: Cycle detected`.

**Comment by jamesnetherton** [comment]
I don't think we need to create `MicrometerRegistryProviderBuildItem`, we only need to consume them.

So I was thinking something like:

```java
@BuildStep
void configureDefaultRegistry(
        List<MicrometerRegistryProviderBuildItem> providers, 
        RootMeterRegistryBuildItem registry,
        CamelMicrometerRecorder recorder) {
    if (providers.isEmpty()) {
        // Adds SimpleMeterRegistry to the Quarkus composite one
        recorder.configureDefaultRegistry(registry.getValue());
    }
}
```

File: extensions/micrometer/deployment/src/main/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerProcessor.java (hunk 1 above)
Comment (JiriOndrusek) [reply extensions/micrometer/deployment/src/main/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerProcessor.java]: N/A

File: extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java (hunk 3 above)
Comment (JiriOndrusek) [reply extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java]: N/A

File: extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java (hunk 2 above)
Comment (JiriOndrusek) [reply extensions/micrometer/deployment/src/test/java/org/apache/camel/quarkus/component/micrometer/deployment/MicrometerSimpleMeterRegistryTest.java]: N/A

**Comment by JiriOndrusek** [comment]
@jamesnetherton your suggestion work as expected. The PR now contains a different implementation of the fix. I added 1 line into the doc, describing autocreation of `SimpleMeterRegistry`.

**Pull request merged**